pkg/apis/authentication/types.go

```diff
@@ -88,3 +89,42 @@ type UserInfo struct {
 
 // ExtraValue masks the value so protobuf can generate
 type ExtraValue []string
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TokenRequest requests a token for a given service account.
+type TokenRequest struct {
+	metav1.TypeMeta
+	// ObjectMeta fulfills the metav1.ObjectMetaAccessor interface so that the stock
+	// REST handler paths work
+	metav1.ObjectMeta
+
+	Spec   TokenRequestSpec
+	Status TokenRequestStatus
+}
+
+// TokenRequestSpec contains client provided parameters of a token request.
+type TokenRequestSpec struct {
+	// Audiences are the intendend audiences of the token. A recipient of a
+	// token must identitfy themself with an identifier in the list of
+	// audiences of the token, and otherwise should reject the token. A
+	// token issued for multiple audiences may be used to authenticate
+	// against any of the audiences listed but implies a high degree of
+	// trust between the target audiences.
+	Audiences []string
+
+	// ValidityDuration is the requested duration of validity of the request. The
+	// token issuer may return a token with a different validity duration so a
+	// client needs to check the 'expiration' field in a response.
+	ValidityDuration metav1.Duration
```

Comment:
I'm not sure how to cleanly represent in a TokenRequest that a token should never expire since I would like to default this field. Can we either:
* use an annotation?
* set a validity duration of 1000 years?

Reply:
Hrm, definitely have never had this before.  We could use the pointer for defaulting, but I don't think a negative duration is appropriate, and MaxInt duration is always ugly.  If we set a duration of zero as never expire and leave it unset, would that cover it?

Reply:
MaxInt is the openssl x509 approach to non-expiring. I dislike the use of sentinel values, but I think it's better then a NeverExpires bool.

Reply:
collapsing the unexpiring case to "really really long expiration" does make the API simpler, if a little weird, and still maintains the possibility for a requester to ask for a 1000 year token and be given a 1 month one (reflected in status).

it removes the watershed division between the two types of tokens, which would make it hard to use a different signing key for eternal tokens, which @mikedanese had mentioned in passing

Reply:
So the docs will say "if you don't want it to expire, set a really really high value, like a billion"?

I suspect people will just pick arbitrary values and set them (no one actually goes and looks up maxint).  I guess that's fine, although it's not really an intent.  I don't want to do IntOrString though.

In the future, if we want to tie the lifetime of the token to a pod, we wouldn't specify that via this API (it would be on the volume source).  We could say:

```
type ValidityType string
const Forever ValidityType = "Forever"
const TimeBound ValidityType = "TimeBound"
const ServerPreferred ValidityType = "ServerPreferred"

type TokenRequestSpec struct {
  Valid ValidityType
  ValiditySeconds *int64
}
```

I wouldn't expect us to use duration here, we should be using seconds.

Reply:
Using service account ID tokens as API access tokens is an abuse. Conflating these two ties our hands with respect to key rotation. If we want to support API access tokens, we should support it as a separate construct. They can still be tied to a service account but we should not tie them to a key that we want to rotate.

My goal here is to support the previous flow, while weening clients onto short lived ID tokens allowing operators to tighten max validity duration of both ID tokens and signing keys. We shouldn't advertise "if you don't want it to expire, set a really really high value, like a billion" in docs. Kubelet should just do it until old clients that depend on this behavior are no longer supported.

If we include "Never" as a valid expiration type clusters will either:
* Never rotate their ID token signing keys
* Not support "Never" tokens, and the API is not portable

Reply:
Ok.  Then `ExpirationSeconds *int64` with an arbitrary high value (our service account tokens today are 1 year).